File: src/main/scala/com/github/hokutomc/lib/scala/nbt/HT_NBTEvidence.scala
package com.github.hokutomc.lib.scala.nbt

import com.github.hokutomc.lib.nbt.HT_NBTUtil
import com.github.hokutomc.lib.scala.HT_Predef._
import com.github.hokutomc.lib.scala.HT_ScalaConversion._
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.BlockPos

/**
 * Created by user on 2015/04/12.
 */
object HT_NBTEvidence {

  trait Compound[E] extends HT_NBTEvidence[E] {

    def writeComp(tag: NBTTagCompound, value: E): Unit

    def readComp(tag: NBTTagCompound): E

    final def write(key: String, tag: NBTTagCompound, a: E): Unit = {
      writeComp(tag.getCompoundTag(key), a)
    }

    final def read(key: String, tag: NBTTagCompound): E = {
      readComp(tag.getCompoundTag(key))
    }
  }


  class EvItemStackArray(val size: Int) extends HT_NBTEvidence.Compound[Array[ItemStack]] {

    override def writeComp(tag: NBTTagCompound, value: Array[ItemStack]): Unit = HT_NBTUtil.writeItemStacks(tag, value)

    override def readComp(tag: NBTTagCompound): Array[ItemStack] = HT_NBTUtil.readItemStacks(tag, size)
  }

  object EvBoolean extends HT_NBTEvidence[Boolean] {
    override def write(key: String, tag: NBTTagCompound, a: Boolean): Unit = tag.setBoolean(key, a)

    override def read(key: String, tag: NBTTagCompound): Boolean = tag.getBoolean(key)
  }

  object EvByte extends HT_NBTEvidence[Byte] {
    override def write(key: String, tag: NBTTagCompound, a: Byte): Unit = tag.setByte(key, a)

    override def read(key: String, tag: NBTTagCompound): Byte = tag.getByte(key)
  }

  object EvByteArray extends HT_NBTEvidence[Array[Byte]] {
    override def write(key: String, tag: NBTTagCompound, a: Array[Byte]): Unit = tag.setByteArray(key, a)

    override def read(key: String, tag: NBTTagCompound): Array[Byte] = tag.getByteArray(key)
  }

  object EvShort extends HT_NBTEvidence[Short] {
    override def write(key: String, tag: NBTTagCompound, a: Short): Unit = tag.setShort(key, a)

    override def read(key: String, tag: NBTTagCompound): Short = tag.getShort(key)
  }

  object EvInt extends HT_NBTEvidence[Int] {
    override def write(key: String, tag: NBTTagCompound, a: Int): Unit = tag.setInteger(key, a)

    override def read(key: String, tag: NBTTagCompound): Int = tag.getInteger(key)
  }

  object EvIntArray extends HT_NBTEvidence[Array[Int]] {
    override def write(key: String, tag: NBTTagCompound, a: Array[Int]): Unit = tag.setIntArray(key, a)

    override def read(key: String, tag: NBTTagCompound): Array[Int] = tag.getIntArray(key)
  }

  object EvLong extends HT_NBTEvidence[Long] {
    override def write(key: String, tag: NBTTagCompound, a: Long): Unit = tag.setLong(key, a)

    override def read(key: String, tag: NBTTagCompound): Long = tag.getLong(key)
  }

  object EvFloat extends HT_NBTEvidence[Float] {
    override def write(key: String, tag: NBTTagCompound, a: Float): Unit = tag.setFloat(key, a)

    override def read(key: String, tag: NBTTagCompound): Float = tag.getFloat(key)
  }

  object EvDouble extends HT_NBTEvidence[Double] {
    override def write(key: String, tag: NBTTagCompound, a: Double): Unit = tag.setDouble(key, a)

    override def read(key: String, tag: NBTTagCompound): Double = tag.getDouble(key)
  }

  object EvString extends HT_NBTEvidence[String] {
    override def write(key: String, tag: NBTTagCompound, a: String): Unit = tag.setString(key, a)

    override def read(key: String, tag: NBTTagCompound): String = tag.getString(key)
  }

  object EvItemStack extends HT_NBTEvidence.Compound[ItemStack] {
    override def writeComp(tag: NBTTagCompound, value: ItemStack): Unit = value.wrapped.writeToNBT(tag)

    override def readComp(tag: NBTTagCompound): ItemStack = ItemStack.loadItemStackFromNBT(tag)
  }

  object EvBlockPos extends HT_NBTEvidence.Compound[BlockPos] {
    override def writeComp(tag: NBTTagCompound, value: BlockPos): Unit = {
      tag.update[Int]("x", value.getX)
      tag.update[Int]("y", value.getY)
      tag.update[Int]("z", value.getZ)
    }

    override def readComp(tag: NBTTagCompound): BlockPos = (for {
      x <- tag[Int]("x")
      y <- tag[Int]("y")
      z <- tag[Int]("z")
    } yield new BlockPos(x, y, z)) orNull
  }

}

trait HT_NBTEvidence[E] {
  def write(key: String, tag: NBTTagCompound, a: E): Unit

  def read(key: String, tag: NBTTagCompound): E
}

File: src/main/java/com/github/hokutomc/lib/nbt/HT_NBTEvidence.java
package com.github.hokutomc.lib.nbt;

import net.minecraft.nbt.NBTTagCompound;

/**
 * Created by user on 2015/06/10.
 */
public abstract class HT_NBTEvidence<T> {
    public abstract void write (String key, NBTTagCompound tagCompound, T value);

    public abstract T read (String key, NBTTagCompound tagCompound);

    public abstract String getTypeString ();


    public static final HT_NBTEvidence<Integer> INT = new HT_NBTEvidence<Integer>() {
        @Override
        public void write (String key, NBTTagCompound tagCompound, Integer value) {
            tagCompound.setInteger(key, value);
        }

        @Override
        public Integer read (String key, NBTTagCompound tagCompound) {
            return tagCompound.getInteger(key);
        }

        @Override
        public String getTypeString () {
            return "int";
        }
    };

    public static final HT_NBTEvidence<Short> SHORT = new HT_NBTEvidence<Short>() {
        @Override
        public void write (String key, NBTTagCompound tagCompound, Short value) {
            tagCompound.setShort(key, value);
        }

        @Override
        public Short read (String key, NBTTagCompound tagCompound) {
            return tagCompound.getShort(key);
        }

        @Override
        public String getTypeString () {
            return "short";
        }
    };

    public static final HT_NBTEvidence<Byte> BYTE = new HT_NBTEvidence<Byte>() {
        @Override
        public void write (String key, NBTTagCompound tagCompound, Byte value) {
            tagCompound.setByte(key, value);
        }

        @Override
        public Byte read (String key, NBTTagCompound tagCompound) {
            return tagCompound.getByte(key);
        }

        @Override
        public String getTypeString () {
            return "byte";
        }
    };

    public static final HT_NBTEvidence<Long> LONG = new HT_NBTEvidence<Long>() {
        @Override
        public void write (String key, NBTTagCompound tagCompound, Long value) {
            tagCompound.setLong(key, value);
        }

        @Override
        public Long read (String key, NBTTagCompound tagCompound) {
            return tagCompound.getLong(key);
        }

        @Override
        public String getTypeString () {
            return "long";
        }
    };

    public static final HT_NBTEvidence<String> STRING = new HT_NBTEvidence<String>() {
        @Override
        public void write (String key, NBTTagCompound tagCompound, String value) {
            tagCompound.setString(key, value);
        }

        @Override
        public String read (String key, NBTTagCompound tagCompound) {
            return tagCompound.getString(key);
        }

        @Override
        public String getTypeString () {
            return "String";
        }
    };

    public static final HT_NBTEvidence<Boolean> BOOLEAN = new HT_NBTEvidence<Boolean>() {
        @Override
        public void write (String key, NBTTagCompound tagCompound, Boolean value) {
            tagCompound.setBoolean(key, value);
        }

        @Override
        public Boolean read (String key, NBTTagCompound tagCompound) {
            return tagCompound.getBoolean(key);
        }

        @Override
        public String getTypeString () {
            return "boolean";
        }
    };

}
